lightning/src/ln/channelmanager.rs

```diff
@@ -10326,18 +10335,24 @@ where
 					Err(()) => Err(InvoiceError::from_string("Unrecognized invoice".to_owned())),
 				};
 
-				match result {
-					Ok(()) => ResponseInstruction::NoResponse,
-					Err(e) => match responder {
-						Some(responder) => responder.respond(OffersMessage::InvoiceError(e)),
-						None => {
-							log_trace!(self.logger, "No reply path for sending invoice error: {:?}", e);
-							ResponseInstruction::NoResponse
-						},
+				match (responder, result) {
+					(Some(responder), Err(e)) => {
+						abandon_if_payment(offers_data.payment_id);
+						responder.respond(OffersMessage::InvoiceError(e))
 					},
+					(None, Err(_)) => {
+						abandon_if_payment(offers_data.payment_id);
+						log_trace!(
+							self.logger,
+							"A response was generated, but there is no reply_path specified for sending the response."
+						);
+						return ResponseInstruction::NoResponse;
+					}
+					_ => return ResponseInstruction::NoResponse,
 				}
 			},
 			OffersMessage::InvoiceError(invoice_error) => {
+				abandon_if_payment(offers_data.payment_id);
```

Comment:
Could you update/add a test to `offers_tests.rs` that show that the payment is abandoned when this case is hit?

Reply:
Could you also use `expect_recent_payment` to show it is in `RecentPaymentDetails::Abandoned`? We should check these for all tests that result in this state.

Reply:
Thanks for your suggestion!
Using `expect_recent_payment` sounds like a good idea, but it might not be very straightforward in this case!

Once `abandon_payment` is called, it removes the payment from `pending_outbound_payments` after generating the appropriate event. Here's the relevant [code snippet](https://github.com/lightningdevkit/rust-lightning/blob/87fc324c71500f017adf535316465825efe083f1/lightning/src/ln/outbound_payment.rs#L1771-L1776).

Given this, checking for the `InvoiceRequestFailed` event might be more fitting to confirm the payment failure.

Additionally, since `abandon_payment` generates an event, any test triggering this should have an unhandled pending event, leading to a failure. This actually happened with `fails_sending_invoice_without_blinded_payment_paths_for_offer` before adding the event handling.

The good news is that since no other tests are failing, it looks like no other tests are reaching this state.

Let me know what you think!

Reply:
Ah, gotcha.